### ISSUE TITLE
plaintext passwords

### ISSUE THREAD

**Issue by nlandolfi** [elos/x#24]

**Comment by viggyfresh** [comment]
Thought - `bcrypt` might be better. Specifically, it handles the randomized salting and retrieval of salting, and is production ready (tons of people use `bcrypt`). Package appears to be well maintained, and usage is absurdly simple. See https://godoc.org/golang.org/x/crypto/bcrypt and https://gowebexamples.github.io/password-hashing/ and http://stackoverflow.com/questions/23259586/bcrypt-password-hashing-in-golang-compatible-with-node-js.

Other options:
* https://github.com/hlandau/passlib - modular, but not super maintained

**Comment by nlandolfi** [comment]
Yeah, I think bcrypt is one such password scheme that uses salting. So under the hood 
```go
func Crypt(pass string) (toStore string) {
    // blah blah 
    bytes, err := bcrypt.GenerateFromPassword([]byte(password), 14)
    // perhaps we also need to return an error here?
    return string(bytes)
}
```

The point being, all of elos doesn't need to know the encryption password used, the interface for the rest of elos is a function that gets the password and a function that checks it...

**Comment by viggyfresh** [comment]
Additionally, late late night reading suggested that hash + salt is insufficient protection these days due to GPU-aided frackers and rainbow tables. The only viable options are bcrypt, scrypt, and something PBK2

**Comment by nlandolfi** [comment]
Ok, I like bcrypt. Especially since it's `golang.org/x`. 

I am a fan of fewer dependencies, and those that we do have begin as close to stdlib as possible.

**Comment by viggyfresh** [comment]
Strongly agree. Hate dependency soup.

**Comment by nlandolfi** [comment]
Ok, `bcrypt`, `pbkdf2`, and `scrypt` are all implemented as subpackages of the `golang.org/x/crypto` path. Therefore I would guess they are all solid options. I am not qualified to make a decision between which algorithm to use...

I'm thinking we could change `auth.Crypt` to `auth.Hash`...

**Comment by viggyfresh** [comment]
`bcrypt` is my vote. `scrypt` is newer but has shown to be vulnerable to GPU-based attacks, `pdbkf2` is approved by NIST (which is a pro but also a con), and `bcrypt` is standard - we can just up the "work" factor from 10 (default) to 20 (for extra security).

**Comment by nlandolfi** [comment]
What's the con for NIST approval? Larger visibility?

**Comment by viggyfresh** [comment]
so the NIST approval is mainly based on the fact that `pdkbf2` is entirely based on already NIST-approved algos (the `SHA` family). If you are a conspiracy theorist...the rest of this thought completes itself. Additionally, `bcrypt` has been around since the late 90s (my CS 140 prof David Mazieres wrote the paper that proposed it), and no one has found exploits against it because it is designed to be a slow hash function that degrades as processors get faster

**Comment by nlandolfi** [comment]
Ok gotcha. I think bcrypt makes sense.
